README.md
DateTime-Set
============

Datetime sets and set math

Changes
Changes for DateTime::Set

dev 2020-10-10
- add set_locale(), following the DateTime API
  Adjusted t/15time_zone.t and docs.

0.3900 2016-10-09
- require Params::Validate

0.3800 2016-05-23
- doc patch, by Lucas Kanashiro.

0.3700 2016-05-17
- end() / max() and start() / min() are aliases to the same method in all classes.
  Added docs and tests. Reported by Vincent Berger.

0.3600 2015-11-11
- move t/22intersects.t to DT::Event::Recurrence module

0.3500 2015-11-10
- [rt.cpan.org #108633] Recurrent event does not intersect indefinite future,
  test contributed by Brett Watson.
- documentation - intersects() returns 0 for false, and 'undef' for
  undecidable.

0.3400 2014-02-12
- documentation and packaging fixes - by David Steinbrunner.
- version number using 4 digits - requested by David Marshall.

0.33   2013-09-15
- bugfix in SpanSet->grep. Reported by Andreas Isberg

0.32   2013-08-27
- ignore duration signal in DateTime::Span->from_datetime_and_duration()
  and use the 'end'/'start' parameters as a cue for the time direction,
  suggested by David Pottage
- more tests of intersections with open/closed ended spans

0.30   2011-04-11
- new test file t/21from_recurrence.t 
  Contributed by Olivier Mengue

0.29   2011-04-01 
- new method is_empty_set - bug #50750

0.28   2009-07-19
- optimized DateTime::Set->as_list().
  This works around a segfault reported by Nils Grunwald.

0.27   2009-04-02
- fixed DateTime::Spanset duration() method.
  Reported by Anton Berezin <tobez@tobez.org>

0.26   2008-12-25
- fixed DateTime::Spanset current() and set_time_zone() methods. 
  Report and tests by Elliot Shank.

0.25   2005-10-05
- t/15time_zone.t does it's own time zone "add_duration" handling

0.24   2005-10-03
- _recurrence.pm warned when the recurrence didn't have occurrences before
  a given start date. Reported by Mark D. Anderson. 

0.23   2005-10-03
- _recurrence.pm died when the recurrence didn't have occurrences before
  a given start date. Reported by Mark D. Anderson. 

0.22   2005-05-06
- DateTime::Set 0.21 dies when as_list is asked to produce a list from
  outside the range of the set. Patch contributed by Stephen Gowing.

0.21   2005-04-06
- no hard limit in count() and as_list(); removed the warnings from the docs.

0.20   2005-02-28
- changed tests to use en_US instead of pt_BR (which changes every year) 
  by Dave Rolsky.
- optimized SpanSet methods for special cases:
  start_set()
  end_set()
  contains( $dt )
  intersects( $dt )
- added an example to count(), by David Nicol.
- added a note about how the result of min()/max() is just a copy of the 
  actual set boundary.  Reported by Ron Hill.

0.19    2004-11-29
- added more comments on as_list() and count() behaviour on large sets.
  Reported by Rick Frankel (rt ticket 8611)

0.18    2004-11-25
- ignore "__DIE__" signal while calculating durations.
  Reported by michelle.agnew (rt ticket 5434)

0.17    2004-08-17
- bugfix: it was trying to set the time zone of an "undef" value; 
  this has broken some DT::Event::Recurrence tests - reported by Bill Moseley

0.1603  2004-07-05
- revised documentation, some examples were rewritten.
- deprecate "iterate" method
- Build.PM requires Set::Infinite 0.59 - reported by Tim Muller-Seydlitz

0.1602  2004-07-02
- fixed time zone handling in the methods:
  set_time_zone
  next/previous/closest/current

- removed limitation of duration size in methods:
  add/add_duration
  subtract/subtract_duration

- requires Set::Infinite 0.59 - uses iterate() experimental 
  argument "backtrack_callback"

- the following methods are now mutators:
  set_time_zone( $tz )
  set( locale => $locale )
  add( unit => $n )
  add_duration( $dur )
  subtract( unit => $n )
  subtract_duration( $dur )

- fixed max/min of an infinite SpanSet
- added test for infinite duration

- from_recurrence() code cleanup

- parameter checking: The 'dates' argument to from_datetimes() 
  must be a list of DateTime objects

0.1601  2004-03-30
- more parameter checking;
  Sets should interoperate better with objects of 
  type DateTime::Event::* and DateTime::Incomplete

0.16  2004-03-29
- Makefile.PL skips processing of "Build.PL"
  Reported by Reinhold May

0.15  2004-03-28
  
  [ ENHANCEMENTS from version 0.14 ]

- Both "Makefile.PL" and "Build.PL" are provided.

- New methods:
  DateTime::SpanSet::current
  DateTime::Spanset::closest
  DateTime::Set::map / grep
  DateTime::SpanSet::map / grep
  DateTime::SpanSet::intersected_spans
  DateTime::SpanSet::start_set / end_set 

- Recurrences can be bounded sets.

- Recurrences can be as small as 1 nanosecond.

- DateTime::SpanSet::duration may return an 'infinite' 
  DateTime::Duration (instead of scalar infinity).

  [ *** BACKWARDS INCOMPATIBILITIES from version 0.14 *** ]

- The next/previous functions used in DT::Set->from_recurrence()
  MUST work with DateTime::Infinite::Future
  and DateTime::Infinite::Past parameters. 

  Failing to test for "infinity" values is known
  to cause an infinite loop in some modules.

- DateTime::Set and DateTime::SpanSet methods do not mutate the set.
  This was the behaviour in version 0.13

0.1413  2004-03-28
- detect_bounded => 1 is deprecated.

0.1412  2004-03-28
- DateTime::Set->from_recurrence - Bounded recurrences are "optional".
  use " detect_bounded => 1 " to turn them on.
  This fixes the backwards incompatibilities created by version 0.1402.

0.1411  2004-03-27
- Makefile.PL.  Suggested by Randal L. Schwartz.

0.1410  2004-03-26
- Build.PL

0.1409  2004-03-23
- evaluation of bounded recurrences is "lazier".
  (DateTime::Incomplete tests are twice as fast)

0.1408  2004-03-22
- refactored S::I::_recurrence class:
  it doesn't use "current" anymore; 
  small bugfixes & optimizations. 
  ("current" was not documented in DT::Set)
- new test: t/17quantize.t
  Tests a method to make "quantization" (or "discretization")
  of spansets.

0.1407  2004-03-16
- New methods:
  DateTime::SpanSet::current
  DateTime::Spanset::closest


lib/DateTime/Span.pm
# Copyright (c) 2003 Flavio Soibelmann Glock. All rights reserved.
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.

package DateTime::Span;

use strict;

use DateTime::Set;
use DateTime::SpanSet;

use Params::Validate qw( validate SCALAR BOOLEAN OBJECT CODEREF ARRAYREF );
use vars qw( $VERSION );

use constant INFINITY     => DateTime::INFINITY;
use constant NEG_INFINITY => DateTime::NEG_INFINITY;
$VERSION = $DateTime::Set::VERSION;

sub set_time_zone {
    my ( $self, $tz ) = @_;

    $self->{set} = $self->{set}->iterate( 
        sub {
            my %tmp = %{ $_[0]->{list}[0] };
            $tmp{a} = $tmp{a}->clone->set_time_zone( $tz ) if ref $tmp{a};
            $tmp{b} = $tmp{b}->clone->set_time_zone( $tz ) if ref $tmp{b};
            \%tmp;
        }
    );
    return $self;
}

# note: the constructor must clone its DateTime parameters, such that
# the set elements become immutable
sub from_datetimes {
    my $class = shift;
    my %args = validate( @_,
                         { start =>
                           { type => OBJECT,
                             optional => 1,
                           },
                           end =>
                           { type => OBJECT,
                             optional => 1,
                           },
                           after =>
                           { type => OBJECT,
                             optional => 1,
                           },
                           before =>
                           { type => OBJECT,
                             optional => 1,
                           },
                         }
                       );
    my $self = {};
    my $set;

    die "No arguments given to DateTime::Span->from_datetimes\n"
        unless keys %args;

    if ( exists $args{start} && exists $args{after} ) {
        die "Cannot give both start and after arguments to DateTime::Span->from_datetimes\n";
    }
    if ( exists $args{end} && exists $args{before} ) {
        die "Cannot give both end and before arguments to DateTime::Span->from_datetimes\n";
    }

    my ( $start, $open_start, $end, $open_end );
    ( $start, $open_start ) = ( NEG_INFINITY,  0 );
    ( $start, $open_start ) = ( $args{start},  0 ) if exists $args{start};
    ( $start, $open_start ) = ( $args{after},  1 ) if exists $args{after};
    ( $end,   $open_end   ) = ( INFINITY,      0 );
    ( $end,   $open_end   ) = ( $args{end},    0 ) if exists $args{end};
    ( $end,   $open_end   ) = ( $args{before}, 1 ) if exists $args{before};

    if ( $start > $end ) {
        die "Span cannot start after the end in DateTime::Span->from_datetimes\n";
    }
    $set = Set::Infinite::_recurrence->new( $start, $end );
    if ( $start != $end ) {
        # remove start, such that we have ">" instead of ">="
        $set = $set->complement( $start ) if $open_start;  
        # remove end, such that we have "<" instead of "<="
        $set = $set->complement( $end )   if $open_end;    
    }

    $self->{set} = $set;
    bless $self, $class;
    return $self;
}

sub from_datetime_and_duration {
    my $class = shift;
    my %args = @_;

    my $key;
    my $dt;
    # extract datetime parameters
    for ( qw( start end before after ) ) {
        if ( exists $args{$_} ) {
           $key = $_;
           $dt = delete $args{$_};
       }
    }

    # extract duration parameters
    my $dt_duration;
    if ( exists $args{duration} ) {
        $dt_duration = $args{duration};
    }
    else {
        $dt_duration = DateTime::Duration->new( %args );
    }
    # warn "Creating span from $key => ".$dt->datetime." and $dt_duration";
    my $other_date;
    my $other_key;
    if ( $dt_duration->is_positive ) {
        if ( $key eq 'end' || $key eq 'before' ) {
            $other_key = 'start';
            $other_date = $dt->clone->subtract_duration( $dt_duration );
        }
        else {
            $other_key = 'before';
            $other_date = $dt->clone->add_duration( $dt_duration );
        }
    }
    else {
        if ( $key eq 'end' || $key eq 'before' ) {
            $other_key = 'start';
            $other_date = $dt->clone->add_duration( $dt_duration );
        }
        else {
            $other_key = 'before';
            $other_date = $dt->clone->subtract_duration( $dt_duration );
        }
    }
    # warn "Creating span from $key => ".$dt->datetime." and ".$other_date->datetime;
    return $class->new( $key => $dt, $other_key => $other_date ); 
}

# This method is intentionally not documented.  It's really only for
# use by ::Set and ::SpanSet's as_list() and iterator() methods.
sub new {
    my $class = shift;
    my %args = @_;

    # If we find anything _not_ appropriate for from_datetimes, we
    # assume it must be for durations, and call this constructor.
    # This way, we don't need to hardcode the DateTime::Duration
    # parameters.
    foreach ( keys %args )
    {
        return $class->from_datetime_and_duration(%args)
            unless /^(?:before|after|start|end)$/;
    }

    return $class->from_datetimes(%args);
}

sub is_empty_set {
    my $set = $_[0];
    $set->{set}->is_null;
}

sub clone { 
    bless { 
        set => $_[0]->{set}->copy,
        }, ref $_[0];
}

# Set::Infinite methods

sub intersection {
    my ($set1, $set2) = @_;
    my $class = ref($set1);
    my $tmp = {};  # $class->new();
    $set2 = $set2->as_spanset
        if $set2->can( 'as_spanset' );
    $set2 = $set2->as_set
        if $set2->can( 'as_set' );
    $set2 = DateTime::Set->from_datetimes( dates => [ $set2 ] ) 
        unless $set2->can( 'union' );
    $tmp->{set} = $set1->{set}->intersection( $set2->{set} );

    # intersection() can generate something more complex than a span.
    bless $tmp, 'DateTime::SpanSet';

    return $tmp;
}

sub intersects {
    my ($set1, $set2) = @_;
    my $class = ref($set1);
    $set2 = $set2->as_spanset
        if $set2->can( 'as_spanset' );
    $set2 = $set2->as_set
        if $set2->can( 'as_set' );
    $set2 = DateTime::Set->from_datetimes( dates => [ $set2 ] ) 
        unless $set2->can( 'union' );
    return $set1->{set}->intersects( $set2->{set} );
}

sub contains {
    my ($set1, $set2) = @_;
    my $class = ref($set1);
    $set2 = $set2->as_spanset
        if $set2->can( 'as_spanset' );
    $set2 = $set2->as_set
        if $set2->can( 'as_set' );
    $set2 = DateTime::Set->from_datetimes( dates => [ $set2 ] ) 
        unless $set2->can( 'union' );
    return $set1->{set}->contains( $set2->{set} );
}

sub union {
    my ($set1, $set2) = @_;
    my $class = ref($set1);
    my $tmp = {};   # $class->new();
    $set2 = $set2->as_spanset
        if $set2->can( 'as_spanset' );
    $set2 = $set2->as_set
        if $set2->can( 'as_set' );
    $set2 = DateTime::Set->from_datetimes( dates => [ $set2 ] ) 
        unless $set2->can( 'union' );
    $tmp->{set} = $set1->{set}->union( $set2->{set} );
 
    # union() can generate something more complex than a span.
    bless $tmp, 'DateTime::SpanSet';

    # # We have to check it's internal structure to find out.
    # if ( $#{ $tmp->{set}->{list} } != 0 ) {
    #    bless $tmp, 'Date::SpanSet';
    # }

    return $tmp;
}

sub complement {
    my ($set1, $set2) = @_;
    my $class = ref($set1);
    my $tmp = {};   # $class->new;
    if (defined $set2) {
        $set2 = $set2->as_spanset
            if $set2->can( 'as_spanset' );
        $set2 = $set2->as_set
            if $set2->can( 'as_set' );
        $set2 = DateTime::Set->from_datetimes( dates => [ $set2 ] ) 
            unless $set2->can( 'union' );
        $tmp->{set} = $set1->{set}->complement( $set2->{set} );
    }
    else {
        $tmp->{set} = $set1->{set}->complement;
    }

    # complement() can generate something more complex than a span.
    bless $tmp, 'DateTime::SpanSet';

    # # We have to check it's internal structure to find out.
    # if ( $#{ $tmp->{set}->{list} } != 0 ) {
    #    bless $tmp, 'Date::SpanSet';
    # }

    return $tmp;
}

sub start { 
    return DateTime::Set::_fix_datetime( $_[0]->{set}->min );
}

*min = \&start;

sub end { 
    return DateTime::Set::_fix_datetime( $_[0]->{set}->max );
}

*max = \&end;

sub start_is_open {
    # min_a returns info about the set boundary 
    my ($min, $open) = $_[0]->{set}->min_a;
    return $open;
}

sub start_is_closed { $_[0]->start_is_open ? 0 : 1 }

sub end_is_open {
    # max_a returns info about the set boundary 
    my ($max, $open) = $_[0]->{set}->max_a;
    return $open;
}

sub end_is_closed { $_[0]->end_is_open ? 0 : 1 }


# span == $self
sub span { @_ }

sub duration { 
    my $dur;

    local $@;
    eval {
        local $SIG{__DIE__};   # don't want to trap this (rt ticket 5434)
        $dur = $_[0]->end->subtract_datetime_absolute( $_[0]->start )
    };
    
    return $dur if defined $dur;

    return DateTime::Infinite::Future->new -
           DateTime::Infinite::Past->new;
}
*size = \&duration;

1;

__END__

=head1 NAME

DateTime::Span - Datetime spans

=head1 SYNOPSIS

    use DateTime;
    use DateTime::Span;

    $date1 = DateTime->new( year => 2002, month => 3, day => 11 );
    $date2 = DateTime->new( year => 2003, month => 4, day => 12 );
    $set2 = DateTime::Span->from_datetimes( start => $date1, end => $date2 );
    #  set2 = 2002-03-11 until 2003-04-12

    $set = $set1->union( $set2 );         # like "OR", "insert", "both"
    $set = $set1->complement( $set2 );    # like "delete", "remove"
    $set = $set1->intersection( $set2 );  # like "AND", "while"
    $set = $set1->complement;             # like "NOT", "negate", "invert"

    if ( $set1->intersects( $set2 ) ) { ...  # like "touches", "interferes"
    if ( $set1->contains( $set2 ) ) { ...    # like "is-fully-inside"

    # data extraction 
    $date = $set1->start;           # first date of the span
    $date = $set1->end;             # last date of the span

=head1 DESCRIPTION

C<DateTime::Span> is a module for handling datetime spans, otherwise
known as ranges or periods ("from X to Y, inclusive of all datetimes
in between").

This is different from a C<DateTime::Set>, which is made of individual
datetime points as opposed to a range. There is also a module
C<DateTime::SpanSet> to handle sets of spans.

=head1 METHODS

=over 4

=item * from_datetimes

Creates a new span based on a starting and ending datetime.

A 'closed' span includes its end-dates:

   $span = DateTime::Span->from_datetimes( start => $dt1, end => $dt2 );

An 'open' span does not include its end-dates:

   $span = DateTime::Span->from_datetimes( after => $dt1, before => $dt2 );

A 'semi-open' span includes one of its end-dates:

   $span = DateTime::Span->from_datetimes( start => $dt1, before => $dt2 );
   $span = DateTime::Span->from_datetimes( after => $dt1, end => $dt2 );

A span might have just a starting date, or just an ending date.
These spans end, or start, in an imaginary 'forever' date:

   $span = DateTime::Span->from_datetimes( start => $dt1 );
   $span = DateTime::Span->from_datetimes( end => $dt2 );
   $span = DateTime::Span->from_datetimes( after => $dt1 );
   $span = DateTime::Span->from_datetimes( before => $dt2 );

You cannot give both a "start" and "after" argument, nor can you give
both an "end" and "before" argument.  Either of these conditions will
cause the C<from_datetimes()> method to die.

To summarize, a datetime passed as either "start" or "end" is included
in the span.  A datetime passed as either "after" or "before" is
excluded from the span.

=item * from_datetime_and_duration

Creates a new span.

   $span = DateTime::Span->from_datetime_and_duration( 
       start => $dt1, duration => $dt_dur1 );
   $span = DateTime::Span->from_datetime_and_duration( 
       after => $dt1, hours => 12 );

The new "end of the set" is I<open> by default.

=item * clone

This object method returns a replica of the given object.

=item * set_time_zone( $tz )

This method accepts either a time zone object or a string that can be
passed as the "name" parameter to C<< DateTime::TimeZone->new() >>.
If the new time zone's offset is different from the old time zone,
then the I<local> time is adjusted accordingly.

If the old time zone was a floating time zone, then no adjustments to
the local time are made, except to account for leap seconds.  If the
new time zone is floating, then the I<UTC> time is adjusted in order
to leave the local time untouched.

=item * duration

The total size of the set, as a C<DateTime::Duration> object, or as a
scalar containing infinity.

Also available as C<size()>.

=item * start

=item * end

First or last dates in the span.

It is possible that the return value from these methods may be a
C<DateTime::Infinite::Future> or a C<DateTime::Infinite::Past>xs object.

If the set ends C<before> a date C<$dt>, it returns C<$dt>. Note that
in this case C<$dt> is not a set element - but it is a set boundary.

=cut

# scalar containing either negative infinity
# or positive infinity.

=item * start_is_closed

=item * end_is_closed

Returns true if the first or last dates belong to the span ( start <= x <= end ).

=item * start_is_open

=item * end_is_open

Returns true if the first or last dates are excluded from the span ( start < x < end ).

=item * union

=item * intersection

=item * complement

Set operations may be performed not only with C<DateTime::Span>
objects, but also with C<DateTime::Set> and C<DateTime::SpanSet>
objects.  These set operations always return a C<DateTime::SpanSet>
object.

    $set = $span->union( $set2 );         # like "OR", "insert", "both"
    $set = $span->complement( $set2 );    # like "delete", "remove"
    $set = $span->intersection( $set2 );  # like "AND", "while"
    $set = $span->complement;             # like "NOT", "negate", "invert"

=item * intersects

=item * contains

These set functions return a boolean value.

    if ( $span->intersects( $set2 ) ) { ...  # like "touches", "interferes"
    if ( $span->contains( $dt ) ) { ...    # like "is-fully-inside"

These methods can accept a C<DateTime>, C<DateTime::Set>,
C<DateTime::Span>, or C<DateTime::SpanSet> object as an argument.

=back

=head1 SUPPORT

Support is offered through the C<datetime@perl.org> mailing list.

Please report bugs using rt.cpan.org

=head1 AUTHOR

Flavio Soibelmann Glock <fglock@gmail.com>

The API was developed together with Dave Rolsky and the DateTime Community.

=head1 COPYRIGHT

Copyright (c) 2003-2006 Flavio Soibelmann Glock. All rights reserved.
This program is free software; you can distribute it and/or modify it
under the same terms as Perl itself.

The full text of the license can be found in the LICENSE file
included with this module.

=head1 SEE ALSO

Set::Infinite

For details on the Perl DateTime Suite project please see
L<http://datetime.perl.org>.

=cut

